code/modules/reagents/machinery/centrifuge.dm

```diff
@@ -0,0 +1,300 @@
+			. = TRUE //no afterattack
+			var/obj/item/weapon/reagent_containers/B = I
+			if(!user.unEquip(B, src))
+				return
+			if(!mainBeaker)
+				mainBeaker = B
+			else
+				separationBeakers.Add(B)
+			to_chat(user, SPAN_NOTICE("You add [B] to [src]."))
+			SSnano.update_uis(src)
+			update_icon()
+			return
+	return ..()
+
+/obj/machinery/centrifuge/MouseDrop_T(atom/C, mob/user, src_location, over_location, src_control, over_control, params)
+	if(!on && istype(C, /obj/item/weapon/reagent_containers) && C.is_open_container())
+		if (!mainBeaker || separationBeakers.len < beakerSlots)
+			if(insert_item(C, user))
+				if(!mainBeaker)
+					mainBeaker = C
+				else
+					separationBeakers.Add(C)
+				to_chat(user, SPAN_NOTICE("You add [C] to [src]."))
+				SSnano.update_uis(src)
+				update_icon()
+	else
+		return ..()
+
+/obj/machinery/centrifuge/on_deconstruction()
+	on = FALSE
+	if(mainBeaker)
+		mainBeaker.forceMove(get_turf(src))
+		mainBeaker = null
+	for(var/obj/item/I in separationBeakers)
+		I.forceMove(get_turf(src))
+	separationBeakers = list()
+	..()
+
+
+/obj/machinery/centrifuge/attack_hand(mob/user)
+	if(..())
+		return TRUE
+
+	user.set_machine(src)
+	ui_interact(user)
+
+/obj/machinery/centrifuge/ui_interact(mob/user, ui_key = "main", datum/nanoui/ui = null, force_open = NANOUI_FOCUS)
+	var/list/data = ui_data()
+
+	// update the ui if it exists, returns null if no ui is passed/found
+	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if (!ui)
+		// the ui does not exist, so we'll create a new() one
+        // for a list of parameters and their descriptions see the code docs in \code\modules\nano\nanoui.dm
+		ui = new(user, src, ui_key, "centrifuge.tmpl", name, 800, 700)
+		// when the ui is first opened this is the data it will use
+		ui.set_initial_data(data)
+		// open the new ui window
+		ui.open()
+
+
+/obj/machinery/centrifuge/ui_data()
+	var/data = list()
+	data["on"] = on
+	data["cycleTime"] = workTime / 10
+	data["timeLeft"] = round((lastActivation + workTime - world.time) / 10)
+	data["maxBeakers"] = beakerSlots
+	data["UPS"] = unitsPerSec
+	
+	if(mainBeaker)
+		data["mainBeaker"] = mainBeaker.reagents.ui_data()
+	var/list/beakersData = list()
+	for(var/i = 1, i <= beakerSlots, i++)
+		var/list/beakerInfo = list()
+		if(i <= separationBeakers.len)
+			var/obj/item/weapon/reagent_containers/B = separationBeakers[i]
+			if(B && B.reagents)
+				beakerInfo = B.reagents.ui_data()
+		beakerInfo["slot"] = i
+		beakersData.Add(list(beakerInfo))
+	data["beakers"] = beakersData
+	return data
+
+
+/obj/machinery/centrifuge/Topic(href, href_list)
+	if(..())
+		return
+
+	if(href_list["power"] && mainBeaker)
+		on = !on
+		lastActivation = world.time
+		update_icon()
+
+	if(href_list["ejectBeaker"] && !on)
+		if(href_list["ejectBeaker"] == "0")
+			mainBeaker.forceMove(get_turf(src))
+			mainBeaker = null
+		else
+			var/slot = text2num(href_list["ejectBeaker"])
```

Comment:
not needed, its tied to other part of the code and if someone fucks that up they will have hard time debugging it. This way it will runtime and point out the problem.